include/odp/api/spec/stash.h

```diff
@@ -0,0 +1,288 @@
+	 *
+	 *  At least 4 byte object size is always supported.
+	 */
+	uint32_t max_obj_size;
+
+} odp_stash_capability_t;
+
+/**
+ * Stash parameters
+ */
+typedef struct odp_stash_param_t {
+	/** Stash type
+	 *
+	 *  Select type of the stash to be created. The default value is
+	 *  ODP_STASH_TYPE_DEFAULT. Use capability to check if additional types
+	 *  are supported.
+	 */
+	odp_stash_type_t type;
+
+	/** Put operation mode
+	 *
+	 *  The default value is ODP_STASH_OP_MT. Usage of ODP_STASH_OP_ST or
+	 *  ODP_STASH_OP_LOCAL mode may improve performance when applicable.
+	 *  If ODP_STASH_OP_LOCAL is used, it must be set to both put_mode and
+	 *  get_mode.
+	 */
+	odp_stash_op_mode_t put_mode;
+
+	/** Get operation mode
+	 *
+	 *  The default value is ODP_STASH_OP_MT. Usage of ODP_STASH_OP_ST or
+	 *  ODP_STASH_OP_LOCAL mode may improve performance when applicable.
+	 *  If ODP_STASH_OP_LOCAL is used, it must be set to both put_mode and
+	 *  get_mode.
+	 */
+	odp_stash_op_mode_t get_mode;
+
+	/** Maximum number of objects
+	 *
+	 *  This is the maximum number of object handles application will store
+	 *  to the stash. The value must not exceed 'max_num_obj' capability.
+	 */
+	uint64_t num_obj;
+
+	/** Object handle size in bytes
+	 *
+	 *  Application uses this size object handles with put and get
+	 *  operations. Valid values are powers of two (1, 2, 4, 8, ... bytes)
+	 *  and must not exceed 'max_obj_size' capability.
+	 */
+	uint32_t obj_size;
+
+} odp_stash_param_t;
+
+/**
+ * Query stash capabilities
+ *
+ * Outputs capabilities of the given stash type on success.
+ * The default stash type (ODP_STASH_TYPE_DEFAULT) is always supported.
+ * The function returns failure if a type is not supported.
+ *
+ * @param[out] capa   Pointer to capability structure for output
+ * @param      type   Stash type
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_stash_capability(odp_stash_capability_t *capa, odp_stash_type_t type);
+
+/**
+ * Initialize stash params
+ *
+ * Initialize an odp_stash_param_t to its default values for all fields
+ *
+ * @param param   Parameter structure to be initialized
+ */
+void odp_stash_param_init(odp_stash_param_t *param);
+
+/**
+ * Create a stash
+ *
+ * This routine is used to create a stash for object handles. It is optional to
+ * give a name. Names do not have to be unique. However, odp_stash_lookup()
+ * returns only a single matching stash.
+ *
+ * @param name     Name of the stash or NULL. Maximum string length is
+ *                 ODP_STASH_NAME_LEN.
+ * @param param    Stash creation parameters.
+ *
+ * @return Handle of the created stash
+ * @retval ODP_STASH_INVALID  Stash could not be created
+ */
+odp_stash_t odp_stash_create(const char *name, odp_stash_param_t *param);
+
+/**
+ * Destroy a stash
+ *
+ * Destroy a previously created stash. Stash must be empty before it is
+ * destroyed. Results are undefined if an attempt is made to destroy a stash
```

Comment:
I don't mind having this requirement but I don't understand the necessity of it. As I understand it stash stores object handle values. Storage for those is controlled entirely by the platform so it can remove it.

Reply:
I think it's safer to start with strict spec and then relax later if needed. This requirement allows implementation to reserve some resources during put() and free those during get(), without cleanup needed during destroy. Also I'd assume that application would get() all stored handles before exit anyway - so that it can free up resources pointed by those handles.

Reply:
Ack.